src/NServiceBus.Core/Transports/Development/DevelopmentTransportMessagePump.cs

```diff
@@ -0,0 +1,306 @@
+
+                    IDevelopmentTransportTransaction transaction;
+
+                    if (transactionMode != TransportTransactionMode.None)
+                    {
+                        transaction = new DirectoryBasedTransaction(path);
+                    }
+                    else
+                    {
+                        transaction = new NoTransaction();
+                    }
+
+                    transaction.BeginTransaction(filePath);
+
+                    await concurrencyLimiter.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+                    var task = Task.Run(async () =>
+                    {
+                        try
+                        {
+                            await ProcessFile(transaction, nativeMessageId).ConfigureAwait(false);
+
+                            transaction.Complete();
+
+                            receiveCircuitBreaker.Success();
+                        }
+                        catch (Exception ex)
+                        {
+                            await receiveCircuitBreaker.Failure(ex).ConfigureAwait(false);
+                        }
+                        finally
+                        {
+                            concurrencyLimiter.Release();
+                        }
+                    }, cancellationToken);
+
+                    task.ContinueWith(t =>
+                        {
+                            Task toBeRemoved;
+                            runningReceiveTasks.TryRemove(t, out toBeRemoved);
+                        }, TaskContinuationOptions.ExecuteSynchronously)
+                        .Ignore();
+
+                    runningReceiveTasks.AddOrUpdate(task, task, (k, v) => task)
+                        .Ignore();
+                }
+
+                if (!filesFound)
+                {
+                    await Task.Delay(10, cancellationToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        async Task ProcessFile(IDevelopmentTransportTransaction transaction, string messageId)
+        {
+            try
+            {
+                var message = File.ReadAllLines(transaction.FileToProcess);
+                var bodyPath = message.First();
+                var headers = HeaderSerializer.Deserialize(string.Join("", message.Skip(1)));
+
+                string ttbrString;
+
+                if (headers.TryGetValue(Headers.TimeToBeReceived, out ttbrString))
+                {
+                    var ttbr = TimeSpan.Parse(ttbrString);
+                    var sentTime = File.GetCreationTimeUtc(transaction.FileToProcess); //file.move preserves create time
+
+                    if (sentTime + ttbr < DateTime.UtcNow)
+                    {
+                        transaction.Commit();
+                        return;
+                    }
+                }
+                var tokenSource = new CancellationTokenSource();
+
+                using (var bodyStream = new FileStream(bodyPath, FileMode.Open))
```

Comment:
Should we use the async mode here?

new FileStream(bodyPath, FileMode.Open, FileAccess.Read, FileShare.None, 4096, FileOptions.Asynchronous)

potentially also `FileOptions.SequentialScan `?

We could also use the `FileOptions.DeleteOnClose` with `Transaction.None`?